proxy/proxy_retries_test.go

```diff
@@ -0,0 +1,228 @@
+			1,
+			0,
+		},
+		{
+			"error response (read failure), don't retry",
+			idempotentQuery,
+			&message.ReadFailure{
+				ErrorMessage: "",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     2,
+				BlockFor:     2,
+				NumFailures:  1,
+			},
+			1,
+			0,
+		},
+		{
+			"error response (write failure), don't retry",
+			idempotentQuery,
+			&message.WriteFailure{
+				ErrorMessage: "",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     2,
+				BlockFor:     2,
+				NumFailures:  1,
+				WriteType:    primitive.WriteTypeSimple,
+			},
+			1,
+			0,
+		},
+		{
+			"unavailable error, retry on the next node",
+			idempotentQuery,
+			&message.Unavailable{
+				ErrorMessage: "Unavailable",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Required:     2,
+				Alive:        1,
+			},
+			2,
+			1,
+		},
+		{
+			"unavailable error w/ non-idempotent query, retry on the next node (same)",
+			nonIdempotentQuery,
+			&message.Unavailable{
+				ErrorMessage: "Unavailable",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Required:     2,
+				Alive:        1,
+			},
+			2,
+			1,
+		},
+		{
+			"read timeout error, retry once on the same node",
+			idempotentQuery,
+			&message.ReadTimeout{
+				ErrorMessage: "ReadTimeout",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     3,
+				BlockFor:     2,
+				DataPresent:  false, // Data wasn't present, read repair, retry
+			},
+			1,
+			1,
+		},
+		{
+			"read timeout error w/ non-idempotent query, retry once on the same node (same)",
+			nonIdempotentQuery,
+			&message.ReadTimeout{
+				ErrorMessage: "ReadTimeout",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     3,
+				BlockFor:     2,
+				DataPresent:  false, // Data wasn't present, read repair, retry
+			},
+			1,
+			1,
+		},
+		{
+			"read timeout error w/ unmet conditions, don't retry",
+			idempotentQuery,
+			&message.ReadTimeout{
+				ErrorMessage: "ReadTimeout",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     2,
+				BlockFor:     2,
+				DataPresent:  true, // Data was present don't retry
+			},
+			1,
+			0,
+		},
+		{
+			"write timeout error, retry once if logged batch",
+			idempotentQuery, // Not a logged batch, but it doesn't matter for this test
+			&message.WriteTimeout{
+				ErrorMessage: "WriteTimeout",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     1,
+				BlockFor:     2,
+				WriteType:    primitive.WriteTypeBatchLog, // Retry if a logged batch
+			},
+			1,
+			1,
+		},
+		{
+			"write timeout error w/ unmet conditions, don't retry",
+			idempotentQuery,
+			&message.WriteTimeout{
+				ErrorMessage: "WriteTimeout",
+				Consistency:  primitive.ConsistencyLevelQuorum,
+				Received:     1,
+				BlockFor:     2,
+				WriteType:    primitive.WriteTypeSimple, // Don't retry for anything other than logged batches
+			},
+			1,
+			0,
+		},
```

Comment:
I'm replicating the behavior of the cpp/java driver default retry policies. I didn't dig into the specific reasons why read/write failures are not retried in the same way as timeouts. I don't fully understand the situations when read/write failures occur, but timeout usually happen when replicas are not responding (possibly because they're dead) and because the default retry policy has a conservative approach of only retrying when a follow up request is likely to succeed maybe that doesn't make sense for failures?